### PR TITLE
Refine tree implementation

### PR DESCRIPTION
**Pull request by jsbean**
@brianlheim:

Here's another shot at it.

### CODE REVIEW

File: Collections/Tree.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/// Node in tree structure
+public enum Tree <T> {
+    
+    // MARK: - Cases
+    
+    /// Empty tree.
+    case empty
+    
+    /// Node with 0 or more subtrees.
+    indirect case node (T, [Tree])
```
Comment (mossheim) [inline Collections/Tree.swift:18]: is it problematic that Tree doesn't have a type specification here?

Comment (jsbean) [reply Collections/Tree.swift:18]: Naw, happy as a clam :).

Comment (jsbean) [reply Collections/Tree.swift:18]: I was wondering that myself, actually. `T` specification is inherited.

Comment (jsbean) [reply Collections/Tree.swift:18]: What is interesting though, is:

```Swift
enum Tree {
    case node (T,  [Tree])
}
```

throws up on itself until the stack blows up, but 

```Swift
enum Tree {
    case node ([Tree])
}
```
is perfectly fine.

Hmmm.

**Pull request closed without merging**